raiden/smart_contracts/TokenNetwork.sol

```diff
@@ -0,0 +1,568 @@
+    {
+        Channel storage channel = channels[channel_identifier];
+
+        require(channel.participants[participant].initialized);
+        require(nonce > channel.participants[participant].nonce);
+        // Transfers can have 0 value
+        require(transferred_amount >= channel.participants[participant].transferred_amount);
+        require(locksroot != channel.participants[participant].locksroot);
+
+        // Update the partner's structure with the data provided
+        // by the closing participant.
+        channel.participants[participant].nonce = nonce;
+        channel.participants[participant].locksroot = locksroot;
+        channel.participants[participant].transferred_amount = transferred_amount;
+    }
+
+    function recoverAddressFromSignature(
+        uint channel_identifier,
+        uint64 nonce,
+        uint256 transferred_amount,
+        bytes32 locksroot,
+        bytes32 additional_hash,
+        bytes signature
+    )
+        view
+        internal
+        returns (address)
+    {
+        require(signature.length == 65);
+
+        bytes32 signed_hash = keccak256(
+            nonce,
+            transferred_amount,
+            locksroot,
+            channel_identifier,
+            address(this),
+            additional_hash
+        );
+
+        var (r, s, v) = signatureSplit(signature);
+        return ecrecover(signed_hash, v, r, s);
+    }
+
+    function signatureSplit(bytes signature)
+        pure
+        internal
+        returns (bytes32 r, bytes32 s, uint8 v)
+    {
+        // The signature format is a compact form of:
+        //   {bytes32 r}{bytes32 s}{uint8 v}
+        // Compact means, uint8 is not padded to 32 bytes.
+        assembly {
+            r := mload(add(signature, 32))
+            s := mload(add(signature, 64))
+            // Here we are loading the last 32 bytes, including 31 bytes
+            // of 's'. There is no 'mload8' to do this.
+            //
+            // 'byte' is not working due to the Solidity parser, so lets
+            // use the second best option, 'and'
+            v := and(mload(add(signature, 65)), 0xff)
+        }
+
+        require(v == 27 || v == 28);
+    }
+
+    // TODO - implement this
+    function encodeLock(
+        uint64 expiration_block,
+        uint amount,
+        bytes32 hashlock)
+        pure
+        internal
+        returns (bytes lock)
+    {
+
+    }
+
+    // TODO - not used anymore now
+    function decodeLock(bytes lock)
+        pure
+        internal
+        returns (uint64 expiration, uint amount, bytes32 hashlock)
+    {
+        require(lock.length == 72);
+
+        // Lock format:
+        // [0:8] expiration
+        // [8:40] amount
+        // [40:72] hashlock
+        assembly {
+            expiration := mload(add(lock, 8))
+            amount := mload(add(lock, 40))
+            hashlock := mload(add(lock, 72))
+        }
+    }
+
+    function computeMerkleRoot(bytes lock, bytes merkle_proof)
```

Comment:
btw, lock can be renamed to `item` or `entry` and this can be moved to another SC if you want.